jobs/build/ocp4/build.groovy

```diff
@@ -0,0 +1,606 @@
+    def full = "${stream}.0"  // default
+    def release = new Date().format("yyyyMMddHHmm")
+
+    def prevBuild = latestOpenshiftRpmBuild(branch)
+    if(params.NEW_VERSION.trim() == "+") {
+        // increment previous build version
+        full = extractBuildVersion(prevBuild)
+        if (!full) { error("Could not determine version from last build '${prevBuild}'") }
+
+        def segments = full.tokenize(".").collect { it.toInteger() }
+        segments[-1]++
+        full = segments.join(".")
+        echo("Using version ${full} incremented from latest openshift package ${prevBuild}")
+    } else if(params.NEW_VERSION) {
+        // explicit version given
+        full = commonlib.standardVersion(params.NEW_VERSION, false)
+        echo("Using NEW_VERSION parameter for version: ${full}")
+    } else if (prevBuild) {
+        // use version from previous build
+        full = extractBuildVersion(prevBuild)
+        if (!full) { error("Could not determine version from last build '${prevBuild}'") }
+        echo("Using version ${full} from latest openshift package ${prevBuild}")
+    }
+
+    if (! full.startsWith("${stream}.")) {
+        // The version we came up with somehow doesn't match what we expect to build; abort
+        error("Determined a version, '${full}', that does not begin with '${stream}.'")
+    }
+
+    stream = stream.tokenize('.').collect { it.toInteger() }
+    return [
+        major: stream[0],
+        minor: stream[1],
+        full: full,
+        release: release,
+    ]
+}
+
+/**
+ * Plan what will be built.
+ * Figure out whether we're building RPMs and/or images, and which ones, based on
+ * the parameters and which sources have changed (if relevant).
+ * Fills in the "buildPlan" property for later stages to use.
+ * @return map which is the buildPlan property of this build.
+ */
+def planBuilds() {
+    if (buildPlan.forceBuild) {
+        echo "Building all as requested regardless of whether source has changed."
+        currentBuild.description = "Force building (whether source changed or not).\n"
+        if (!buildPlan.buildRpms) {
+            currentBuild.description += "RPMs: not building.\n"
+        }
+        currentBuild.description += "Will create RPM compose.\n"
+        if (!buildPlan.buildImages) {
+            currentBuild.description += "Images: not building.\n"
+        }
+        return buildPlan
+    }
+
+    // otherwise we need to scan sources.
+    echo "Building only where source has changed."
+    currentBuild.description = "Building sources that have changed.\n"
+
+    def changed = [:]
+    try {
+        def yamlData = readYaml text: buildlib.doozer(
+            """
+            ${doozerOpts}
+            ${includeExclude "rpms", buildPlan.rpmsIncluded, buildPlan.rpmsExcluded}
+            ${includeExclude "images", buildPlan.imagesIncluded, buildPlan.imagesExcluded}
+            config:scan-sources --yaml
+            """, [capture: true]
```

Comment:
You're depending on doozer never accidentally spitting out cruft text to `stdout` here. I recall a recent conversation where this happened, doozer spit out some garbage into the yaml stream. I'd suggest to consider parsing from the output yaml file but it looks like doozer doesn't give the choice to `-o something.yaml` yet. I guess this is just something to keep in mind.

Reply:
Yes. If doozer is broken it might do that even when it's supposed to output yaml. I was thinking one of our automated doozer tests should try a few commands that have easily-defined output and ensure that's not happening.

Reply:
Although I suppose `-o foo.yaml` would be a more robust solution.